docs/source/rfc/on-chain-scripting.org

```diff
@@ -0,0 +1,954 @@
+4. Rust core team is contributing to =wasmtime=
+5. It might be required for bridging Substrate based networks
+6. We have experience in it within our team and teams closely working to us
+7. Developed by wide community
+8. Execution can be easily limited with `fuel` (approximate of number
+   of assembly instructions that any script is allowed to execute at
+   most)
+
+Point 2 has questionable relevance for blockchain smart contract
+development.  The only relation is half-tangential in that web
+developers (who primarily use other languages), represent a large
+demographic of current programmers.
+
+Point 3 is misleading. 32-bit WASM is a Tier 2 target, which to quote
+the [[https://doc.rust-lang.org/nightly/rustc/platform-support.html][original documentation]],
+
+#+BEGIN_EXAMPLE
+Tier 2 targets can be thought of as "guaranteed to build". The Rust
+project builds official binary releases for each tier 2 target, and
+automated builds ensure that each tier 2 target builds after each
+change. Automated tests are not always run so it's not guaranteed to
+produce a working build, but tier 2 targets often work to quite a good
+degree and patches are always welcome! For the full requirements, see
+Tier 2 target policy in the Target Tier Policy.
+#+END_EXAMPLE
+
+in other words, it is one entry in a list of 50 architectures,
+including ones like =aarch64-apple-tvos=, =bpfeb-unknown-none=, which
+was /not/ mentioned in the eBPF discussion, =loongarch-*= etc.
+
+Point 4, is a logical fallacy. Being good at one thing doesn't
+translate into being good at another unrelated thing. It's like
+arguing that coffee that was brewed by Picasso is somehow better than
+coffee brewed by someone else.
+
+Point 5, neglects to consider the overall architecture of Iroha, and
+that even if Iroha uses WASM, because those smart contracts use a
+completely different API and ABI and are not modular like pallets in
+parity substrate, that then the smart contracts for all intents and
+purposes can be considered binary black boxes, which is also true of
+any other standard.
+
+Point 6 did not last. Most specialists could equally quickly learn to
+work with eBPF and with x86 binaries as well.
+
+Point 7 is a good indicator of health of the project, not of its
+suitability for purpose. The community is largely interested in using
+WASM to execute logic on the browser that might otherwise require
+native support from the architecture.
+
+Point 8 is an extremely minor one, because despite having =fuel= as a
+concept, it is not used to limit execution, but instead the even
+parity substrate decides weights via a benchmarking process.
+
+The cons are equally underdeveloped.
+
+1. Requires nightly* for the smart contracts themselves
+2. Difficult to limit memory further than 3GB in =wasmtime= (needs more
+   research)
+3. No Interface Types Support - so it has to be simulated through
+   passing serialized bytes or mapping memory
```

Comment:
Btw, I think there _is_ support for interface types (now called wasm [Component Model](https://github.com/WebAssembly/component-model)). One project I know of using it is [Ambient](https://github.com/AmbientRun/Ambient/tree/main/crates/wasm)